kernel/memory/src/paging/mapper.rs

```diff
@@ -867,3 +867,104 @@ pub fn mapped_pages_unmap(
 pub fn mapper_from_current() -> Mapper {
     Mapper::from_current()
 }
+
+
+/// An immutably borrowed [`MappedPages`] object that derefs to `&T`.
+///
+/// When dropped, the borrow ends and the contained `MappedPages` is dropped and unmapped.
+pub struct BorrowedMappedPages<T: FromBytes> {
+    ptr: NonNull<T>,
+    mp: MappedPages,
+}
+impl<T: FromBytes> Deref for BorrowedMappedPages<T> {
+    type Target = T;
+    fn deref(&self) -> &T {
+        // SAFETY:
+        // ✅ The pointer is properly aligned because we use `NonNull`.
+        // ✅ The pointer is dereferenceable, as it has been bounds checked by `MappedPages::as_type()`.
+        // ✅ The pointer has been initialized in the constructor `try_into_borrowed()`.
+        // ✅ The lifetime of the returned reference `&T` is tied to the lifetime of the `MappedPages`,
+        //    ensuring that the `MappedPages` object will persist at least as long as the reference.
+        unsafe { self.ptr.as_ref() }
+    }
+}
+impl<T: FromBytes> BorrowedMappedPages<T> {
+    /// Immutably borrows the given `MappedPages` as an instance of type `&T` 
+    /// starting at the given `offset` into the `MappedPages`.
+    ///
+    /// See [`MappedPages::as_type()`] for more info.
+    ///
+    /// Returns an error containing the unmodified `MappedPages` and a string
+    /// describing the error.
+    pub fn try_into_borrowed(
+        mp: MappedPages,
+        offset: usize,
+    ) -> Result<BorrowedMappedPages<T>, (MappedPages, &'static str)> {
+        let mut borrowed_mp = BorrowedMappedPages {
+            ptr: NonNull::<T>::dangling(),
+            mp,
+        };
+        borrowed_mp.ptr = match borrowed_mp.mp.as_type::<T>(offset) {
```

Comment:
This is an issue with `as_type`, but I can't comment on its source code as it's not part of the PR.

`as_type` doesn't check alignment which I believe is unsound. The [`zerocopy::FromBytes`](https://docs.rs/zerocopy/latest/zerocopy/trait.FromBytes.html) documentation states that:

> `FromBytes` types can safely be deserialized from an untrusted sequence of bytes because any byte sequence corresponds to a valid instance of the type.

But it doesn't mention anything about alignment. However, the existence of the [`zerocopy::Unalign`](https://docs.rs/zerocopy/latest/zerocopy/struct.Unalign.html) type suggests that if we aren't using `zerocopy::Unalign` we should be checking pointer alignment.

Reply:
Fixed